drivers/net/bnxt/bnxt_ethdev.c

```diff
@@ -1115,27 +1116,68 @@ static int bnxt_set_vf_vlan_filter_op(struct rte_eth_dev *dev, uint16_t vlan,
 				uint64_t vf_mask, uint8_t vlan_on)
 {
 	struct bnxt *bp = (struct bnxt *)dev->data->dev_private;
-	int i;
-	int ret;
+	int i, j;
 	int rc = 0;
+	int dflt_vnic;
+	struct bnxt_vnic_info vnic;
+	struct bnxt_vlan_table_entry *ve;
 
 	if (!bp->pf.vf_info)
 		return -EINVAL;
 
 	for (i=0; vf_mask; i++, vf_mask >>= 1) {
 		if (vf_mask & 1) {
-			bp->pf.vf_info[i].dflt_vlan = vlan_on ? vlan : 0;
-			if (bnxt_hwrm_func_qcfg_current_vf_vlan(bp, i) == bp->pf.vf_info[i].dflt_vlan)
+			if (bp->pf.vf_info[i].vlan_table == NULL)
 				continue;
-			ret = bnxt_hwrm_set_vf_vlan(bp, i);
-			if (ret)
-				rc = ret;
+			if (vlan_on) {
+				/* First, search for a duplicate... */
+				for (j=0; j<bp->pf.vf_info[i].vlan_count; j++) {
+					if (bp->pf.vf_info[i].vlan_table[j].vid == vlan)
+						continue;
+				}
+				if (j < bp->pf.vf_info[i].vlan_count)
+					continue;
+
+				/* Now check that there's space */
+				if (bp->pf.vf_info[i].vlan_count == getpagesize() / sizeof(struct bnxt_vlan_table_entry)) {
+					RTE_LOG(ERR, PMD, "VF %d VLAN table is full, cannot add VLAN %u\n", i, vlan);
+					continue;
+				}
+
+				/* And finally, add to the end of the table */
+				ve = &bp->pf.vf_info[i].vlan_table[bp->pf.vf_info[i].vlan_count++];
+				ve->tpid = rte_cpu_to_be_16(0x8100);	// TODO: Hardcoded TPID
+				ve->vid = vlan;
+			}
+			else {
+				for (j=0; j<bp->pf.vf_info[i].vlan_count; j++) {
+					if (bp->pf.vf_info[i].vlan_table[j].vid != vlan)
+						continue;
+					memmove(&bp->pf.vf_info[i].vlan_table[j], &bp->pf.vf_info[i].vlan_table[j+1],
+					    getpagesize() - ((j+1) * sizeof(struct bnxt_vlan_table_entry)));
+					j--;
+				}
+			}
+			dflt_vnic = bnxt_hwrm_func_qcfg_vf_dflt_vnic_id(bp, i);
+			if (dflt_vnic < 0) {
+				RTE_LOG(ERR, PMD, "Unable to get default VNIC for VF %d\n", i);
+			}
+			else {
+				if (bnxt_hwrm_vnic_qcfg(bp, &vnic, dflt_vnic) == 0) {
+					bnxt_hwrm_cfa_l2_set_rx_mask(bp, &vnic, bp->pf.vf_info[i].vlan_count, bp->pf.vf_info[i].vlan_table);
```

Comment:
I'm really not sure what we could expect the caller to do with an error code though.  We can't roll back, and we can't communicate the new state.

I suppose the caller could exit...